docs/design/enums_or_ids.md

```diff
@@ -0,0 +1,140 @@
+Enums or Identifiers
+====================
+
+An *entity* is a discrete item in a set. For example, English is an entity: it is a specific instance of a language.
+
+There are two general ways to represent entities: *enumerations* and *identifiers*. An *enumeration* is a fixed, finite list of entities. An *identifier* is a universally understood space of numbers or strings, which may or may not be finite. For example, the identifier `"en"` is a language code according to a standard published by ISO.
+
+This design doc helps explain how we think about entities in ICU4X and when to use enums versus identifiers to represent them.
+
+## Decision Tree
```

Comment:
To my mind, the key difference between enumerations and the new-type pattern is that Rust enums come with a guarantee that the value you're looking at comes from a known universe of values. If you have an enum-typed value, then Rust guarantees that it is valid and the rest of the codebase understands how to deal with it. The newtype case does not have the same guarantee.

This is both good and bad. The good part, as we discussed in the meeting, is that it lets older code work with newer data. The bad part is that it makes it harder for the compiler to help you. For example, if I accidentally write `Script::Laitn`, it's a compiler error; if I accidentally write `Script("Laitn")`, it's a bold new development in linguistics. Similarly, if I have a `#[non_exhaustive]` enum, I can still depend on the compiler to verify that matches _inside_ the crate are exhaustive, whereas the newtype case requires a wildcard pattern even in our own code. 

(From Shane's response to Zibi:
> I don't find the "attention has to be continuously paid" argument to be compelling. The backing data store of an identifier is private, and all constructors on the newtype should perform validation or be named unchecked. I don't see that as being a challenge.

The difference between enums and newtypes here is that Rust won't let you construct an enum without validation (or `unsafe`), whereas rules about newtype constructors can only be enforced by human vigilance.)

On this basis, I think that enums are more ergonomic / less error-prone. That's not to say that I think we should use enums everywhere; as I mentioned in the meeting last week, I find the old-code/new-data argument compelling. But I think we should make sure we are getting something by giving up ergonomics.

All of that is to say: I think this decision tree leans a little too heavily towards identifiers. For example, `GeneralCategory` has more than 20 possible values, but no new values will ever be added, so there's no reason not to make it an enum. I would flip the questions around:

1. How often are new entities added to the set?
   - Set is fixed or grows rarely: Use an enum.
   - Spontaneously: Use a newtype.
   - By an authority, at least once every few years: Continue to the next question.

2. How consistent is the handling of each entity? How hard is it to write code to handle unknown entities? Do we need to know anything about the entity in our own code, or can we pass it straight through to the data provider?
   - Very consistent: Use a newtype.
   - Inconsistent: Use an enum. 

Note that a Script newtype doesn't do anything with our current implementation of enumerated properties, because script data is accessed via ResourceKey (eg "sc=Latn"), and there's no ResourceKey defined for scripts we don't know about. If we want to be able to handle new Script data with old code, then we need to overhaul the data provider.

Reply:
I agree with this. I would add on to the note at the end that planning to support the representation of new unknown values doesn't necessarily confer any new abilities (ex: as you said, can't load UnicodeSet data for new property values we haven't seen). But we can at least prevent blowing up in the use case of returning property values where, for example, the value we deserialize from the CodePointTrie of newer data is outside the defined set of variants in our older enum.

Maybe what that means is that avoiding that ungraceful error state is the limit of what we can guarantee as the benefit here for identifiers over enums?

Reply:
One more data point: I'm currently working on converting the existing Script enum into an identifier and it's ... kind of unpleasant?

In particular, we run into this problem I mentioned above:
> Similarly, if I have a #[non_exhaustive] enum, I can still depend on the compiler to verify that matches inside the crate are exhaustive, whereas the newtype case requires a wildcard pattern even in our own code.

If `Script` is just `pub struct Script(pub(crate) u16)` with a bunch of associated constants, then the match in [`get_script_val_set`](https://github.com/unicode-org/icu4x/blob/adf02273511847c7b3422a0916e05d8ac7c236ec/components/uniset/src/props.rs#L550-L556) is no longer exhaustive, so we have to figure out what to do if we're given a `Script` wrapping an identifier we don't know about. Meanwhile, to create the associated constants, I've ended up writing this, which feels like I'm holding the hammer backwards:
```

Reply:
For example, we don't have constants for all the language codes we support. You create then via their language code as a string.

Why should scripts be different than language codes in this regard?

Reply:
In other words, in cases where it makes sense to use identifiers, there shouldn't necessarily be a comprehensive list of all of the identifiers in code.  All that we need is a data-driven map from the 4-letter codes to the numbers, which is backed by the data provider.  ICU4C updates its header files to add in the new entries with each release; I would like if we didn't have to do that and could instead take all mutable data and stick it in the data provider.

Reply:
I think two-letter language codes have more developer mindshare as a canonical identifier than 4-letter script codes. For example, the ECMA spec makes both the long and short names available for scripts, but defines the _long_ name as the "canonical property value". People are less likely to have the 4-letter code memorized, so it's nice if there's a convenient list of human-readable values (which can also be used in eg IDE autocompletion).

Also, the implicit values (Common, Inherited, Unknown) that developers are most likely to need to hardcode have particularly awful names (Zyyy, Zinh/Qaai, Zzzz). As a developer, I would much prefer to write (or read) `Script::Common` rather than `Script::from_code("Zyyy").unwrap()`.

Basically, I think we provide a better developer experience if there's a list of identifiers available, and I'm not convinced that having to recompile for a new release is a sufficient problem. Given that ICU4C currently doesn't support swapping out data without recompiling, nobody's existing workflow will be blocked if ICU4X doesn't support that use case.

Reply:
> Also, the implicit values (Common, Inherited, Unknown) that developers are most likely to need to hardcode have particularly awful names (Zyyy, Zinh/Qaai, Zzzz). As a developer, I would much prefer to write (or read) Script::Common rather than Script::from_code("Zyyy").unwrap().

Actually, this might be a good question for the checklist: "Do some values have a different meaning / require special handling? If yes, lean towards enum." Country codes are a flat list, with no special cases; scripts have special cases, which makes it more likely that people will need to refer to those special cases directly, which makes the ergonomic issues more salient.

Reply:
> I think two-letter language codes have more developer mindshare as a canonical identifier than 4-letter script codes.

Not sure if I agree with that.  Anyone familiar with BCP-47 already knows the 4-letter script codes.  "zh-Hans" versus "zh-Hant", for instance.  I find `script!("Hant")` *at least as* readable as `Script::TraditionalHan`.

> Actually, this might be a good question for the checklist: "Do some values have a different meaning / require special handling? If yes, lean towards enum." Country codes are a flat list, with no special cases; scripts have special cases, which makes it more likely that people will need to refer to those special cases directly, which makes the ergonomic issues more salient.

In my mental model, when identifiers are used, it is totally acceptable and even expected to have a certain subset of commonly used values available as constants.  For scripts, that could be perhaps the "special" script codes (Zyyy, Zzzz, ...) plus Latn (since Latn is often useful in computing applications).

> Also, the implicit values (Common, Inherited, Unknown) that developers are most likely to need to hardcode have particularly awful names (Zyyy, Zinh/Qaai, Zzzz). As a developer, I would much prefer to write (or read) `Script::Common` rather than `Script::from_code("Zyyy").unwrap()`.

I simply don't see the use case behind defining all 200+ script codes as constants *or* enums. If you are a developer who wants to work with a long-tail script code, then it's not difficult to ask that you write out the 4-letter script code.

> I'm not convinced that having to recompile for a new release is a sufficient problem. Given that ICU4C currently doesn't support swapping out data without recompiling, nobody's existing workflow will be blocked if ICU4X doesn't support that use case.

I have been trying to make the case since the inception of ICU4X that a core feature of ICU4X relative to ICU4C is the ability to decouple code version from data version.  We know of several Google clients where this is a major pain point in adopting ICU; they want to have a single data file shared across a whole app or platform, even if different pieces of the app or platform were compiled against different versions of the code.  It would be a *major* shift if ICU4X started imposing the code-data requirement, and I believe it would greatly diminish the value of ICU4X for a number of clients.

Reply:
> Not sure if I agree with that. Anyone familiar with BCP-47 already knows the 4-letter script codes. "zh-Hans" versus "zh-Hant", for instance. I find script!("Hant") at least as readable as Script::TraditionalHan.
Neither "Hans" nor "Hant" is a valid script in this context. Han characters are assigned `Script::Han`, with four-letter code "Hani".

Reply:
> I have been trying to make the case since the inception of ICU4X that a core feature of ICU4X relative to ICU4C is the ability to decouple code version from data version. We know of several Google clients where this is a major pain point in adopting ICU; they want to have a single data file shared across a whole app or platform, even if different pieces of the app or platform were compiled against different versions of the code. It would be a major shift if ICU4X started imposing the code-data requirement, and I believe it would greatly diminish the value of ICU4X for a number of clients.

Fair point. I retract my objection.

> In my mental model, when identifiers are used, it is totally acceptable and even expected to have a certain subset of commonly used values available as constants. For scripts, that could be perhaps the "special" script codes (Zyyy, Zzzz, ...) plus Latn (since Latn is often useful in computing applications).
I simply don't see the use case behind defining all 200+ script codes as constants or enums. If you are a developer who wants to work with a long-tail script code, then it's not difficult to ask that you write out the 4-letter script code.

I guess I see it the other way around: if we're going to make Latin available as a constant / enum value, then I don't see the harm in also making other scripts available. The 4-letter codes are obscure enough (and sufficiently inconsistent with uses elsewhere) that users are generally going to have to go look up the correct code. It doesn't cost us anything to make these values available.

I think it's useful to think about two cases: Script as input to ICU4X, and Script as output from ICU4X.

**Input**: If the user tries to create a Script to pass in to ICU4X, we have to validate it. 

In the identifier case, the user provides a 4-letter code. We look it up in a data-driven map and, if it's a valid script, create a 4-byte TinyStr wrapped in a newtype. With some proc macro trickery, we can do this at compile time for constants. One advantage here is that new scripts can be added to the data post-compilation, and we can still correctly parse the new script name and return the new data. 

In the enum case, constructing a known constant script is trivial. If the user needs to look up a dynamic value (most likely a string) then it needs to be converted to an enum first. We can provide a similar API that accepts 4-letter codes. Relative to identifiers, the disadvantage is that we can't accept codes that weren't known at compile time; a possible advantage is that we're matching against a bunch of known constants, so LLVM may be able to do a better job optimizing.

**Output**: If the user asks "what script does this character belong to", we have to be able to represent the answer.

In the identifier case, we can look up the integer value of the script in the codepoint trie, and then look up that integer value in a data-driven side table to determine the associated 4-letter code. The advantage here is that we can return a script value that was unknown at compile time.

In the enum case, we also look up the integer value of the script in the codepoint trie. If it's a value we recognize, we're done. If not, we have to decide what to do. We have several options here, but I think the best one is to just return `Script::Unknown`. Note that so long as scripts are allocated sequentially, I think this ends up being a single comparison. If the user wants the short name or long name, we can provide separate functions to look those up, equivalent to the side-table lookup that we always do in the identifier case.

So it seems to me that both approaches are capable of handling the code/data separation, and the advantages of the identifier version (better support for eg Klingon in codebases that were compiled before it was added to Unicode) are smaller than the ergonomic advantages of using enums, at least in the specific case of Script where new additions are expected to see incredibly marginal use.

(I'm starting to think there may also be a proc-macro based approach that gives us the data-driven forward-compatibility benefits of identifiers with the ergonomic advantages of enums, but I want to try writing some code before I make strong claims about that.)

Reply:
Fair points.

> I guess I see it the other way around: if we're going to make Latin available as a constant / enum value, then I don't see the harm in also making other scripts available. The 4-letter codes are obscure enough (and sufficiently inconsistent with uses elsewhere) that users are generally going to have to go look up the correct code. It doesn't cost us anything to make these values available.

I see the cost as being the ongoing human cost of having to update ICU4X APIs to conform to the latest Unicode standards; it puts our release process in the way of downstream consumers fetching the latest and greatest Unicode data, instead of allowing them to just re-generate data offline and use it with an older version of ICU4X.

> In the enum case, we also look up the integer value of the script in the codepoint trie. If it's a value we recognize, we're done. If not, we have to decide what to do. We have several options here, but I think the best one is to just return `Script::Unknown`.

This is a possibility.  It still makes the ICU4X dev team the bottleneck, because new APIs are required in order to interpret the new data.

*If* we were to use an enum, I think `Script::Unknown` should be a variant that wraps a `u16`, such that a data-driven map between script numbers and script codes or script names can still work correctly.